storage/innobase/log/log0online.cc

```diff
@@ -44,37 +45,392 @@ Online database log parsing for changed page tracking */
+         type != MLOG_TABLE_DYNAMIC_META
+#ifdef UNIV_LOG_LSN_DEBUG
+         && type != MLOG_LSN
+#endif
+      ;
+}
+
+/* Mutex protecting log_bmp_sys and log buffers there */
+static ib_mutex_t log_bmp_sys_mutex;
+
+/** A redo log byte buffer with associated LSN values */
+template <std::size_t CAPACITY>
+class log_buffer {
+ protected:
+  using buffer_type = std::array<byte, CAPACITY>;
+
+ public:
+  using size_type = typename buffer_type::size_type;
+  using const_iterator = typename buffer_type::const_iterator;
+
+ protected:
+  buffer_type buffer;
+  static const constexpr auto capacity = CAPACITY;
+  size_type current_size{0};
+  lsn_t start_lsn{0};
+  lsn_t current_lsn{0};
+  const_iterator current_ptr{buffer.cbegin()};
+  lsn_t limit_lsn{0};
+
+  log_buffer() noexcept {}
+
+#ifdef UNIV_DEBUG
+  MY_NODISCARD bool invariants() const noexcept {
+    ut_ad(mutex_own(&log_bmp_sys_mutex));
+    ut_ad(start_lsn <= current_lsn);
+    ut_ad(current_lsn <= limit_lsn);
+    ut_ad(static_cast<decltype(start_lsn)>(current_ptr - buffer.cbegin()) <=
+          current_size);
+    return true;
+  }
+#endif
+
+ public:
+  MY_NODISCARD const_iterator ccurrent() const noexcept {
+    ut_ad(mutex_own(&log_bmp_sys_mutex));
+    return current_ptr;
+  }
+
+  void set_limit(lsn_t limit) {
+    ut_ad(mutex_own(&log_bmp_sys_mutex));
+    ut_ad(limit >= limit_lsn);
+    limit_lsn = limit;
+  }
+
+  MY_NODISCARD lsn_t get_current_lsn() const noexcept {
+    ut_ad(mutex_own(&log_bmp_sys_mutex));
+    return current_lsn;
+  }
+};
+
 static const constexpr auto FOLLOW_SCAN_SIZE = 4 * UNIV_PAGE_SIZE_MAX;
 
+static_assert(FOLLOW_SCAN_SIZE % OS_FILE_LOG_BLOCK_SIZE == 0,
+              "FOLLOW_SCAN_SIZE must be a multiple of OS_FILE_LOG_BLOCK_SIZE");
+
+static const constexpr auto LOG_BLOCK_SIZE_NO_TRL =
+    OS_FILE_LOG_BLOCK_SIZE - LOG_BLOCK_TRL_SIZE;
+
+static const constexpr auto LOG_BLOCK_BOUNDARY_LSN_PAD =
+    LOG_BLOCK_HDR_SIZE + LOG_BLOCK_TRL_SIZE;
+
+/** The buffer for reading log data, filled in by recv_read_log_recs and moved
+in chunks to the parse buffer while skipping log block headers and trailers. */
+class log_read_buffer final : public log_buffer<FOLLOW_SCAN_SIZE> {
+ public:
+  void read(lsn_t read_start, lsn_t read_end) noexcept;
+
+  MY_NODISCARD bool is_data_available() const noexcept {
+    ut_ad(mutex_own(&log_bmp_sys_mutex));
+    const auto result = ccurrent() < cend() && get_current_lsn() < limit_lsn;
+    if (result) {
+      ut_ad(invariants());
+    }
+    return result;
+  }
+
+  /** Check the log block checksum.
+  @return true if the log block checksum is OK, false otherwise.  */
+  MY_NODISCARD bool is_current_block_valid() const noexcept;
+
+  MY_NODISCARD uint32_t get_current_block_data_len() const noexcept {
+    ut_ad(is_current_block_valid());
+    return log_block_get_data_len(ccurrent());
+  }
+
+  MY_NODISCARD const_iterator cend() const noexcept {
+    return &buffer[current_size];
```

Comment:
`buffer.cend()` ?

Reply:
might point differently: buffer.cend() is always at FOLLOW_SCAN_SIZE, current_size is set by ::read()

Reply:
Oh, my bad. Missed the `current_size` part.